monitor/service.go

```diff
@@ -0,0 +1,306 @@
+
+		// Start periodic writes to system.
+		s.wg.Add(1)
+		go func() {
+			defer s.wg.Done()
+			tick := time.NewTicker(s.storeInterval)
+			for {
+				select {
+				case <-tick.C:
+					if err := s.storeStatistics(); err != nil {
+						s.Logger.Printf("failed to write statistics to %s: %s", s.storeAddress, err.Error())
+					}
+				case <-s.done:
+					s.Logger.Printf("terminating storage of statistics to %s", s.storeAddress)
+					return
+				}
+
+			}
+		}()
+	}
+
+	// If enabled, expose all expvar data over HTTP.
+	if s.expvarAddress != "" {
+		listener, err := net.Listen("tcp", s.expvarAddress)
+		if err != nil {
+			return err
+		}
+
+		go func() {
+			http.Serve(listener, nil)
+		}()
+		s.Logger.Printf("expvar information available on %s", s.expvarAddress)
+	}
+	return nil
+}
+
+// Close closes the monitor service.
+func (s *Service) Close() {
+	s.Logger.Println("shutting down monitor service")
+	close(s.done)
+	s.wg.Wait()
+	s.done = nil
+}
+
+// SetLogger sets the internal logger to the logger passed in.
+func (s *Service) SetLogger(l *log.Logger) {
+	s.Logger = l
+}
+
+// Register registers a client with the given name and tags.
+func (s *Service) Register(name string, tags map[string]string, client Client) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	c := &clientWithMeta{
+		Client: client,
+		name:   name,
+		tags:   tags,
+	}
+	s.registrations = append(s.registrations, c)
+	s.Logger.Printf(`'%s:%v' registered for monitoring`, name, tags)
+	return nil
+}
+
+// ExecuteStatement executes monitor-related query statements.
+func (s *Service) ExecuteStatement(stmt influxql.Statement) *influxql.Result {
+	switch stmt := stmt.(type) {
+	case *influxql.ShowStatsStatement:
+		return s.executeShowStatistics(stmt)
+	default:
+		panic(fmt.Sprintf("unsupported statement type: %T", stmt))
+	}
+}
+
+// executeShowStatistics returns the statistics of the registered monitor client in
+// the standard form expected by users of the InfluxDB system.
+func (s *Service) executeShowStatistics(q *influxql.ShowStatsStatement) *influxql.Result {
+	stats, _ := s.statistics()
+	rows := make([]*influxql.Row, len(stats))
+
+	for n, stat := range stats {
+		row := &influxql.Row{}
+		values := make([]interface{}, 0, len(stat.Tags)+len(stat.Values))
+
+		row.Columns = append(row.Columns, "name")
+		values = append(values, stat.Name)
+
+		for _, k := range stat.tagNames() {
+			row.Columns = append(row.Columns, k)
+			values = append(values, stat.Tags[k])
+		}
+		for _, k := range stat.valueNames() {
+			row.Columns = append(row.Columns, k)
+			values = append(values, stat.Values[k])
+		}
+		row.Values = [][]interface{}{values}
+		rows[n] = row
+	}
+	return &influxql.Result{Series: rows}
+}
+
+// statistics returns the combined statistics for all registered clients.
+func (s *Service) statistics() ([]*statistic, error) {
+	statistics := make([]*statistic, 0, len(s.registrations))
+	for _, r := range s.registrations {
```

Comment:
Yeah, good catch. I have a mutex already for this object, but forgot to take it here.